### PR TITLE
[Storage] Support Uniform Bucket Level Access

### PR DESCRIPTION
**Pull request by frankyn**
Storage feature request to support the rename of Bucket Policy Only to Uniform Bucket Level Access.

- [x] Update integration tests to use uniform bucket-level access instead.
- [x] Update documentation.
- [x] Review from GCS eng team(@thanhld94).
- [ ] Fix pending for backend bug: 134695162. Remove workaround.

### CODE REVIEW

File: storage/google/cloud/storage/bucket.py [hunk 1]
```diff
@@ -280,6 +280,12 @@ class IAMConfiguration(dict):
...
     :params bucket: Bucket for which this instance is the policy.
 
+    :type uniform_bucket_level_access_enabled: bool
+    :params bucket_policy_only_enabled: (optional) whether the IAM-only policy is enabled for the bucket.
+
+    :type uniform_bucket_level_locked_time: :class:`datetime.datetime`
+    :params uniform_bucket_level_locked_time: (optional) When the bucket's IAM-only policy was ehabled.  This value should normally only be set by the back-end API.
+
     :type bucket_policy_only_enabled: bool
     :params bucket_policy_only_enabled: (optional) whether the IAM-only policy is enabled for the bucket.
```
Comment (tseaver) [inline storage/google/cloud/storage/bucket.py]: Should `bucket_policy_only_enabled` and `bucket_policy_only_locked_time` be documented as deprecated?

File: storage/google/cloud/storage/bucket.py [hunk 2]
```diff
@@ -292,12 +298,32 @@ def __init__(
         bucket,
         bucket_policy_only_enabled=False,
         bucket_policy_only_locked_time=None,
+        uniform_bucket_level_access_enabled=False,
+        uniform_bucket_level_access_locked_time=None,
```
Comment (tseaver) [inline storage/google/cloud/storage/bucket.py]: Ordering should match the order documented above.  The constructor probably needs to check for / resolve conflicts between `uniform_bucket_level_access_enabled` / `bucket_policy_only_enabled` and `uniform_bucket_level_access_locked_time` / `bucket_policy_only_locked_time` first, which would simplify the code setting up `data`.

File: storage/google/cloud/storage/bucket.py [hunk 3]
```diff
@@ -314,6 +340,7 @@ def from_api_repr(cls, resource, bucket):
         :rtype: :class:`IAMConfiguration`
         :returns: Instance created from resource.
         """
+
```
Comment (tseaver) [inline storage/google/cloud/storage/bucket.py]: Stray line.

File: storage/google/cloud/storage/bucket.py [hunk 4]
```diff
@@ -329,36 +356,54 @@ def bucket(self):
...
 
+    @bucket_policy_only_enabled.setter
+    def bucket_policy_only_enabled(self, value):
+        self.uniform_bucket_level_access_enabled(value)
+
     @property
     def bucket_policy_only_locked_time(self):
-        """Deadline for changing :attr:`bucket_policy_only_enabled` from true to false.
+        """Alias for uniform_bucket_level_access."""
+        return self.uniform_bucket_level_access_locked_time()
```
Comment (tseaver) [inline storage/google/cloud/storage/bucket.py]: Don't call the property, just return it, e.g.:
```suggestion
        return self.uniform_bucket_level_access_locked_time
```

File: storage/tests/unit/test_bucket.py [hunk 5]
```diff
@@ -197,8 +197,8 @@ def test_ctor_defaults(self):
         config = self._make_one(bucket)
 
         self.assertIs(config.bucket, bucket)
-        self.assertFalse(config.bucket_policy_only_enabled)
-        self.assertIsNone(config.bucket_policy_only_locked_time)
+        self.assertFalse(config.uniform_bucket_level_access_enabled)
+        self.assertIsNone(config.uniform_bucket_level_access_locked_time)
```
Comment (tseaver) [inline storage/tests/unit/test_bucket.py]: We need to retain tests for the deprecated properties (which would have exposed the "can't call properties" bugs I noted above).

File: storage/tests/unit/test_bucket.py [hunk 6]
```diff
@@ -208,12 +208,12 @@ def test_ctor_explicit(self):
...
         config = self._make_one(
-            bucket, bucket_policy_only_enabled=True, bucket_policy_only_locked_time=now
+            bucket, uniform_bucket_level_access_enabled=True, uniform_bucket_level_access_locked_time=now
         )
 
         self.assertIs(config.bucket, bucket)
-        self.assertTrue(config.bucket_policy_only_enabled)
-        self.assertEqual(config.bucket_policy_only_locked_time, now)
+        self.assertTrue(config.uniform_bucket_level_access_enabled)
+        self.assertEqual(config.uniform_bucket_level_access_locked_time, now)
```
Comment (tseaver) [inline storage/tests/unit/test_bucket.py]: Likewise, we need to continue to assert values for the deprecated properties here.

We also need separate testcases for passing `bucket_policy_only_enabled` / `bucket_policy_only_locked_time` to the constructor.

File: storage/tests/unit/test_bucket.py [hunk 7]
```diff
@@ -223,30 +223,30 @@ def test_from_api_repr_w_empty_resource(self):
...
-        resource = {"bucketPolicyOnly": {}}
+        resource = {"uniformBucketLevelAccess": {}}
 
         config = klass.from_api_repr(resource, bucket)
 
         self.assertIs(config.bucket, bucket)
-        self.assertFalse(config.bucket_policy_only_enabled)
-        self.assertIsNone(config.bucket_policy_only_locked_time)
+        self.assertFalse(config.uniform_bucket_level_access_enabled)
+        self.assertIsNone(config.uniform_bucket_level_access_locked_time)
```
Comment (tseaver) [inline storage/tests/unit/test_bucket.py]: We need to retain the assertions about the deprecated properties here.

Also, do we need a separate testcase for a resource which contains only an empty `bucketPolicyOnly` mapping?  I.e., can we guarantee that the back-end won't return such a resource?

File: storage/tests/unit/test_bucket.py [hunk 8]
```diff
@@ -223,30 +223,30 @@ def test_from_api_repr_w_empty_resource(self):
...
-        resource = {"bucketPolicyOnly": {"enabled": False}}
+        resource = {"uniformBucketLevelAccess": {"enabled": False}}
 
         config = klass.from_api_repr(resource, bucket)
 
         self.assertIs(config.bucket, bucket)
-        self.assertFalse(config.bucket_policy_only_enabled)
-        self.assertIsNone(config.bucket_policy_only_locked_time)
+        self.assertFalse(config.uniform_bucket_level_access_enabled)
+        self.assertIsNone(config.uniform_bucket_level_access_locked_time)
```
Comment (tseaver) [inline storage/tests/unit/test_bucket.py]: Preserve the assertions about the deprecated properties here, too.

Also, do we need a separate testcase for a resource which contains only a populated `bucketPolicyOnly` mapping? I.e., can we guarantee that the back-end won't return such a resource?

File: storage/tests/unit/test_bucket.py [hunk 9]
```diff
@@ -266,16 +266,16 @@ def test_from_api_repr_w_enabled(self):
         config = klass.from_api_repr(resource, bucket)
 
         self.assertIs(config.bucket, bucket)
-        self.assertTrue(config.bucket_policy_only_enabled)
-        self.assertEqual(config.bucket_policy_only_locked_time, now)
+        self.assertTrue(config.uniform_bucket_level_access_enabled)
+        self.assertEqual(config.uniform_bucket_level_access_locked_time, now)
```
Comment (tseaver) [inline storage/tests/unit/test_bucket.py]: Preserve the assertions about the deprecated properties here, too.

Also, do we need a separate testcase for a resource which contains only a populated `bucketPolicyOnly` mapping? I.e., can we guarantee that the back-end won't return such a resource?

File: storage/tests/unit/test_bucket.py [hunk 10]
```diff
@@ -266,16 +266,16 @@ def test_from_api_repr_w_enabled(self):
...
 
-    def test_bucket_policy_only_enabled_setter(self):
+    def test_uniform_bucket_level_access_enabled_setter(self):
```
Comment (tseaver) [inline storage/tests/unit/test_bucket.py]: We need to preserve separate test cases for the deprecated properties' setters.

File: storage/tests/unit/test_bucket.py [hunk 11]
```diff
@@ -266,16 +266,16 @@ def test_from_api_repr_w_enabled(self):
...
         bucket = self._make_bucket()
         config = self._make_one(bucket)
 
-        config.bucket_policy_only_enabled = True
+        config.uniform_bucket_level_access_enabled = True
 
-        self.assertTrue(config["bucketPolicyOnly"]["enabled"])
+        self.assertTrue(config["uniformBucketLevelAccess"]["enabled"])
```
Comment (tseaver) [inline storage/tests/unit/test_bucket.py]: Preserve the assertion about the deprecated prroperty here, too.

**Comment by JesseLovelace** [comment]
Hi @tseaver, thank you for reviewing. Moving forward with this, we actually need you to take over this PR, and I will review it. Can you update it with the changes you've suggested, as well as removing the workaround? (That should work now). I'm working on getting you access to the one-pager in case you need that

**Comment by tseaver** [comment]
Superseded by #9475.

**Pull request closed without merging**